### PR TITLE
Hashable Rules

### PR DESCRIPTION
**Pull request by comane**
Rules (for cuts) are made to tuple instead of list.
This makes them hashable and allows eg them to contained within a `TupleComp` object

### CODE REVIEW

File: validphys2/src/validphys/config.py [hunk 1]
```diff
@@ -1261,7 +1262,7 @@ def parse_filter_rules(self, filter_rules: (list, type(None))):
         """A list of filter rules. See https://docs.nnpdf.science/vp/filters.html
         for details on the syntax"""
         log.warning("Overwriting filter rules")
-        return filter_rules
+        return tuple(frozendict(rule) for rule in filter_rules) if filter_rules else None
```
Comment (scarlehoff) [inline validphys2/src/validphys/config.py:1265]: Please  change the docstr from list to tuple. You might want to mention ad well that each rule is given as an inmutable dictionary.

File: validphys2/src/validphys/config.py [hunk 2]
```diff
@@ -1286,17 +1286,16 @@ def produce_rules(
...
         """Produce filter rules based on the user defined input and defaults."""
-        from validphys.filters import Rule, RuleProcessingError, default_filter_rules_input
 
         theory_parameters = theoryid.get_description()
 
         if filter_rules is None:
             # Don't bother loading the rules if we are not using them.
             if use_cuts is not CutsPolicy.INTERNAL:
-                return None
+                return tuple()
```
Comment (scarlehoff) [inline validphys2/src/validphys/config.py:1298]: Is this change necessary? `None` should be hashable and this could have collateral effects (like some code doing `if x is None`)

File: validphys2/src/validphys/config.py [hunk 3]
```diff
@@ -1359,45 +1358,69 @@ def load_default_default_filter_settings(self, spec):
...
+        return parsed_filter_defaults
 
     def produce_defaults(
         self,
         q2min=None,
         w2min=None,
         maxTau=None,
         default_filter_settings=None,
-        filter_defaults={},
+        filter_defaults=FilterDefaults,
```
Comment (scarlehoff) [inline validphys2/src/validphys/config.py:1392]:
```suggestion
        filter_defaults=None,
```

And check early whether it is None and change it to an empty dict.

File: validphys2/src/validphys/config.py [hunk 4]
```diff
@@ -1359,45 +1358,69 @@ def load_default_default_filter_settings(self, spec):
...
 
-        if q2min is not None and "q2min" in filter_defaults and q2min != filter_defaults["q2min"]:
-            raise ConfigError("q2min defined multiple times with different values")
-        if w2min is not None and "w2min" in filter_defaults and w2min != filter_defaults["w2min"]:
-            raise ConfigError("w2min defined multiple times with different values")
+            if q2min is not None and filter_defaults["q2min"] != q2min:
+                raise ConfigError("q2min defined multiple times with different values")
+            
+            if w2min is not None and filter_defaults["w2min"] != w2min:
+                raise ConfigError("w2min defined multiple times with different values")
```
Comment (scarlehoff) [inline validphys2/src/validphys/config.py:1411]: I think this is changing too much the behaviour of the function. I would leave this as close as possible to the previous behaviour.

File: validphys2/src/validphys/config.py (hunk 2 above)
Comment (comane) [reply validphys2/src/validphys/config.py]: I had some trouble at some point with a bit of the code saying `non Iterable` and therefore changed it to `tuple()`

I now changed it back, let's see after the tests

**Comment by comane** [comment]
> Thanks for this, could you also remove frozenargs from the conda recipe / pyproject? It should not be needed anymore

I am not sure I understand, do you mean `frozendict`?  Because that one is actually still needed

**Comment by scarlehoff** [comment]
> Because that one is actually still needed

I thought it was only needed by the filter rules?

File: validphys2/src/validphys/config.py (hunk 4 above)
Comment (scarlehoff) [reply validphys2/src/validphys/config.py:1413]: No, before it would check whether I'm passing a `filter_defaults` that is incompatible with the values of `w2min`, `q2min` etc that I'm also passing.

Now this check is only done if the `filter_defaults` I'm passing is an instance of `FilterDefaults`. Note that regardless of the default input of the function I can call it with e.g.

`produce_defaults(qmin=1, filter_defaults={"q2min": 4})`

File: validphys2/src/validphys/config.py [hunk 5]
```diff
@@ -1258,10 +1265,10 @@ def load_default_default_filter_rules(self, spec):
             )
 
     def parse_filter_rules(self, filter_rules: (list, type(None))):
-        """A list of filter rules. See https://docs.nnpdf.science/vp/filters.html
-        for details on the syntax"""
+        """A tuple of FilterRule objects. The rules are frozen dataclasses.
```
Comment (scarlehoff) [inline validphys2/src/validphys/config.py:1268]:
```suggestion
        """A tuple of FilterRule objects. Rules are immutable after parsing.
```

File: validphys2/src/validphys/config.py [hunk 6]
```diff
@@ -1271,12 +1278,15 @@ def parse_default_filter_rules_recorded_spec_(self, spec):
         return spec
 
     def parse_added_filter_rules(self, rules: (list, type(None)) = None):
-        return rules
+        """
+        Returns a tuple of AddedFilterRule objects. The rules are frozen dataclasses.
```
Comment (scarlehoff) [inline validphys2/src/validphys/config.py:1282]:
```suggestion
        Returns a tuple of AddedFilterRule objects. Rules are immutable after parsing.
```

File: validphys2/src/validphys/config.py [hunk 7]
```diff
@@ -1271,12 +1278,15 @@ def parse_default_filter_rules_recorded_spec_(self, spec):
...
+        AddedFilterRule objects inherit from FilterRule objects.
```
Comment (scarlehoff) [inline validphys2/src/validphys/config.py:1283]: Why is `AddedFilterRule` not just a `FilterRule`?

File: validphys2/src/validphys/config.py [hunk 8]
```diff
@@ -1305,7 +1314,7 @@ def produce_rules(
         try:
             rule_list = [
                 Rule(
-                    initial_data=rule,
+                    initial_data=rule.to_dict(),
```
Comment (scarlehoff) [inline validphys2/src/validphys/config.py:1317]: I think that instead of passing `.to_dict` here, it is better to do, in the class `Rule`, 

```python
# For compatibility with legacy code that passed a dictionary
# TODO: make sure there are no calls left to `Rules` with a dictionary input
initial_data = initial_data.to_dict()
```

(you might want to cover this with a `isinstance` or `try-except`)

File: validphys2/src/validphys/config.py [hunk 9]
```diff
@@ -1359,45 +1367,71 @@ def load_default_default_filter_settings(self, spec):
...
+            Currently these limits are ``q2min``, ``w2min``, and ``maxTau``.
         """
         log.warning("Overwriting filter defaults")
-        return filter_defaults
+
+        parsed_filter_defaults = FilterDefaults(
+                                            q2min = filter_defaults["q2min"] if "q2min" in filter_defaults else None, 
+                                            w2min = filter_defaults["w2min"] if "w2min" in filter_defaults else None,
+                                            maxTau = filter_defaults["maxTau"] if "maxTau" in filter_defaults else None,
+                                            )
```
Comment (scarlehoff) [inline validphys2/src/validphys/config.py:1392]:
```suggestion
        parsed_filter_defaults = FilterDefaults(**filter_defaults)
```

And then you can initialise the attributes of the class `FilterDefaults` to `None`.

If for some reason that cannot be done, you can always do `filter_defaults.get("q2min")` which is equivalent to doing `filter_defaults["q2min"] if "q2min" in filter_defaults else None`

File: validphys2/src/validphys/filters.py [hunk 10]
```diff
@@ -99,18 +101,63 @@ class FatalRuleError(Exception):
...
+        return filtered_dict
+
+
+@dataclasses.dataclass(frozen=True)
+class AddedFilterRule(FilterRule):
+    """
+    Dataclass which carries extra filter rule that is added to the
+    default rule.
+    """
+    pass
```
Comment (scarlehoff) [inline validphys2/src/validphys/filters.py:144]: as said above, I don't think we should separate AddedFilterRules and FilterRules, they should be at the same level once they are read.

File: validphys2/src/validphys/filters.py [hunk 11]
```diff
@@ -511,7 +558,7 @@ def __init__(self, initial_data: dict, *, defaults: dict, theory_parameters: dic
         self.rule_string = self.rule
         self.defaults = defaults
         self.theory_params = theory_parameters
-        ns = {*self.numpy_functions, *self.defaults, *self.variables, "idat", "central_value"}
+        ns = {*self.numpy_functions, *self.defaults.to_dict().keys(), *self.variables, "idat", "central_value"}
```
Comment (scarlehoff) [inline validphys2/src/validphys/filters.py:561]:
```suggestion
        ns = {*self.numpy_functions, *self.defaults.to_dict(), *self.variables, "idat", "central_value"}
```

Not that it makes a difference though.

File: validphys2/src/validphys/loader.py [hunk 12]
```diff
@@ -674,10 +674,10 @@ def check_default_filter_rules(self, theoryid, defaults=None):
         th_params = theoryid.get_description()
         if defaults is None:
             defaults = default_filter_settings_input()
-        return [
-            Rule(inp, defaults=defaults, theory_parameters=th_params, loader=self)
+        return tuple(
+            Rule(inp.to_dict(), defaults=defaults, theory_parameters=th_params, loader=self)
```
Comment (scarlehoff) [inline validphys2/src/validphys/loader.py:678]:
```suggestion
            Rule(inp, defaults=defaults, theory_parameters=th_params, loader=self)
```

(same as above, I think it'd be better to have `Rule` take the dataclass by default and convert it to a dict inside)

File: validphys2/src/validphys/tests/test_filter_rules.py [hunk 13]
```diff
@@ -120,7 +120,7 @@ def test_added_rules():
             {
                 "speclabel": "fewer data",
                 "added_filter_rules": [
-                    {"dataset": "ATLAS_1JET_8TEV_R06_PTY", "rule": "pT < 1000", "reson": "pt cut"}
+                    {"dataset": "ATLAS_1JET_8TEV_R06_PTY", "rule": "pT < 1000", "reason": "pt cut"}
```
Comment (scarlehoff) [inline validphys2/src/validphys/tests/test_filter_rules.py:123]: We are already fixing bugs and naming problems thanks to these classes! ^__^

File: validphys2/src/validphys/config.py [hunk 14]
```diff
@@ -1271,12 +1278,15 @@ def parse_default_filter_rules_recorded_spec_(self, spec):
...
+        """
+        Returns a tuple of AddedFilterRule objects. The rules are frozen dataclasses.
+        AddedFilterRule objects inherit from FilterRule objects.
+        """
+        return tuple(AddedFilterRule(**rule) for rule in rules) if rules else None
 
     # Every parallel replica triggers a series of calls to this function,
     # which should not happen since the rules are identical among replicas.
     # E.g for NNPDF4.0 with 2 parallel replicas 693 calls, 3 parallel replicas 1001 calls...
-    @freeze_args
```
Comment (scarlehoff) [inline validphys2/src/validphys/config.py:1279]: I think now you can remove the comment above!

File: validphys2/src/validphys/config.py (hunk 7 above)
Comment (comane) [reply validphys2/src/validphys/config.py:1283]: They are the same indeed (Added... inherits from FilterRule). But it might be useful to differentiate between the two for instance for error messages or don't know what.

So I think I prefer having two "different" objects for these

File: validphys2/src/validphys/filters.py [hunk 15]
```diff
@@ -107,12 +107,14 @@ class FilterDefaults:
...
-    maxTau: float
+    q2min: float = None
+    w2min: float = None
+    maxTau: float = None
 
     def to_dict(self):
-        return dataclasses.asdict(self)
+        default_dict = dataclasses.asdict(self)
+        filtered_dict = {k: v for k, v in default_dict.items() if v is not None}
+        return filtered_dict
```
Comment (scarlehoff) [inline validphys2/src/validphys/filters.py:117]: Is this necessary? Before you were also returning `None` when there were no defaults.

File: validphys2/src/validphys/filters.py [hunk 16]
```diff
@@ -457,10 +506,20 @@ class Rule:
 
     numpy_functions = {"sqrt": np.sqrt, "log": np.log, "fabs": np.fabs}
 
-    def __init__(self, initial_data: dict, *, defaults: dict, theory_parameters: dict, loader=None):
+    def __init__(self, initial_data: dataclasses.dataclass, *, defaults: dict, theory_parameters: dict, loader=None):
```
Comment (scarlehoff) [inline validphys2/src/validphys/filters.py:509]:
```suggestion
    def __init__(self, initial_data: FilterRule, *, defaults: dict, theory_parameters: dict, loader=None):
```

And I would be inclined to write in the error below that only `FilterRule`s are allowed.
The fact that we accept `Mapping`s is to make sure we don't break any new code yet but I don't think it should be advertised.

File: validphys2/src/validphys/filters.py (hunk 15 above)
Comment (comane) [reply validphys2/src/validphys/filters.py:117]: yes, you are right. I don't think it is necessary

File: validphys2/src/validphys/config.py (hunk 4 above)
Comment (scarlehoff) [reply validphys2/src/validphys/config.py:1414]: I think this is the only part left to be addressed to avoid having `filter_defaults` which are incompatible with the input values of `maxTau, q2min, w2min`

File: validphys2/src/validphys/config.py [hunk 17]
```diff
@@ -1359,45 +1364,66 @@ def load_default_default_filter_settings(self, spec):
...
+        
         if w2min is not None and "w2min" in filter_defaults and w2min != filter_defaults["w2min"]:
             raise ConfigError("w2min defined multiple times with different values")
 
-        if (
-            maxTau is not None
-            and "maxTau" in filter_defaults
-            and maxTau != filter_defaults["maxTau"]
-        ):
+        if maxTau is not None  and filter_defaults["maxTau"] != maxTau:
```
Comment (scarlehoff) [inline validphys2/src/validphys/config.py:1416]:
```suggestion
        if maxTau is not None and filter_defaults.get("maxTau", maxTau) != maxTau:
```

**Review by scarlehoff — APPROVED** [review]
Seems ok. I've done a few tests and I haven't seen anything obviously wrong.

File: validphys2/src/validphys/filters.py [hunk 18]
```diff
@@ -99,18 +101,63 @@ class FatalRuleError(Exception):
...
+    IC: str = None
+
+    def to_dict(self):
+        rule_dict = dataclasses.asdict(self)
+        filtered_dict = {k: v for k, v in rule_dict.items() if v is not None}
+        return filtered_dict
+
+
+@dataclasses.dataclass(frozen=True)
+class AddedFilterRule(FilterRule):
```
Comment (RoyStegeman) [inline validphys2/src/validphys/filters.py:139]: What is the purpose of this class?

File: validphys2/src/validphys/config.py (hunk 7 above)
Comment (comane) [reply validphys2/src/validphys/config.py:1283]: > What is the purpose of this class?
Hi @RoyStegeman  there was a little discussion here that might answer your question.

Comment (RoyStegeman) [reply validphys2/src/validphys/config.py:1283]: Okay thanks, I wasn't sure it was intentional or forgotten
